File: vx01_ws/src/vx01_control/aerial/vx01_aerial_control/src/aerial_controller_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <std_msgs/msg/bool.hpp>
#include <mavros_msgs/msg/state.hpp>
#include <mavros_msgs/srv/set_mode.hpp>
#include <mavros_msgs/srv/command_bool.hpp>

using namespace std::chrono_literals;

class AerialControllerNode : public rclcpp::Node {
public:
    AerialControllerNode() : Node("aerial_controller_node")
    {
        sub_cmd_ = create_subscription<geometry_msgs::msg::TwistStamped>(
            "/drone/cmd_vel", 10,
            [this](const geometry_msgs::msg::TwistStamped::SharedPtr msg) {
                latest_cmd_          = *msg;
                last_cmd_time_       = now();
                has_received_cmd_    = true;
            });

        sub_state_ = create_subscription<mavros_msgs::msg::State>(
            "/mavros/state", 10,
            [this](const mavros_msgs::msg::State::SharedPtr msg) {
                is_connected_ = msg->connected;
                is_armed_     = msg->armed;
                current_mode_ = msg->mode;
            });

        pub_vel_ = create_publisher<geometry_msgs::msg::TwistStamped>(
            "/mavros/setpoint_velocity/cmd_vel", 10);

        sub_arm_ = create_subscription<std_msgs::msg::Bool>(
            "/drone/arm", 10,
            [this](const std_msgs::msg::Bool::SharedPtr msg) {
                sendArmCommand(msg->data);
            });

        arm_client_  = create_client<mavros_msgs::srv::CommandBool>("/mavros/cmd/arming");
        mode_client_ = create_client<mavros_msgs::srv::SetMode>("/mavros/set_mode");

        heartbeat_timer_ = create_wall_timer(100ms, [this]() { publishSetpoint(); });

        RCLCPP_INFO(get_logger(),
            "AerialController ready. "
            "Publish geometry_msgs/TwistStamped to /drone/cmd_vel for flight. "
            "Publish std_msgs/Bool to /drone/arm to arm (true) or disarm (false).");
    }

private:

    void publishSetpoint()
    {
        if (!is_connected_) return;

        geometry_msgs::msg::TwistStamped cmd;
        cmd.header.stamp    = now();
        cmd.header.frame_id = "base_link";

        if (has_received_cmd_) {
            double age = (now() - last_cmd_time_).seconds();
            if (age < 0.5) {
                cmd.twist = latest_cmd_.twist;
            }
        }

        pub_vel_->publish(cmd);
    }

    void sendArmCommand(bool arm)
    {
        if (!arm_client_->wait_for_service(1s)) {
            RCLCPP_WARN(get_logger(), "Arming service not available");
            return;
        }
        auto req = std::make_shared<mavros_msgs::srv::CommandBool::Request>();
        req->value = arm;
        arm_client_->async_send_request(req,
            [this, arm](rclcpp::Client<mavros_msgs::srv::CommandBool>::SharedFuture future) {
                auto resp = future.get();
                if (resp->success) {
                    RCLCPP_INFO(get_logger(), "%s successful", arm ? "ARM" : "DISARM");
                } else {
                    RCLCPP_WARN(get_logger(), "%s FAILED (result=%d)",
                        arm ? "ARM" : "DISARM", resp->result);
                }
            });
    }

    rclcpp::Subscription<geometry_msgs::msg::TwistStamped>::SharedPtr   sub_cmd_;
    rclcpp::Subscription<mavros_msgs::msg::State>::SharedPtr     sub_state_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr         sub_arm_;
    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr      pub_vel_;
    rclcpp::TimerBase::SharedPtr                                 heartbeat_timer_;
    rclcpp::Client<mavros_msgs::srv::CommandBool>::SharedPtr     arm_client_;
    rclcpp::Client<mavros_msgs::srv::SetMode>::SharedPtr         mode_client_;

    geometry_msgs::msg::TwistStamped latest_cmd_{};
    rclcpp::Time              last_cmd_time_{0, 0, RCL_ROS_TIME};
    bool                      has_received_cmd_ = false;
    bool                      is_connected_     = false;
    bool                      is_armed_         = false;
    std::string               current_mode_     = "STABILIZE";
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AerialControllerNode>());
    rclcpp::shutdown();
    return 0;
}

File: vx01_ws/src/vx01_control/aerial/vx01_aerial_control/vx01_aerial_control/aerial_controller_node.py
"""
aerial_controller_node.py — Velocity passthrough & arm/disarm for VX-01 drone.

Subscribes:
    /drone/cmd_vel   (geometry_msgs/TwistStamped)  — velocity commands
    /drone/arm       (std_msgs/Bool)               — arm (True) / disarm (False)
    /mavros/state    (mavros_msgs/State)            — FCU state

Publishes:
    /mavros/setpoint_velocity/cmd_vel (geometry_msgs/TwistStamped)

Services called:
    /mavros/cmd/arming   (mavros_msgs/CommandBool)
    /mavros/set_mode     (mavros_msgs/SetMode)
"""

import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped
from std_msgs.msg import Bool
from mavros_msgs.msg import State
from mavros_msgs.srv import CommandBool, SetMode


class AerialControllerNode(Node):

    def __init__(self):
        super().__init__('aerial_controller_node')

        # ── State 
        self._connected = False
        self._armed = False
        self._mode = 'STABILIZE'
        self._latest_cmd = TwistStamped()
        self._last_cmd_time = self.get_clock().now()
        self._has_cmd = False

        # ── Subscribers 
        self.create_subscription(
            TwistStamped, '/drone/cmd_vel', self._cmd_vel_cb, 10)
        self.create_subscription(
            State, '/mavros/state', self._state_cb, 10)
        self.create_subscription(
            Bool, '/drone/arm', self._arm_cb, 10)

        # ── Publisher 
        self._vel_pub = self.create_publisher(
            TwistStamped, '/mavros/setpoint_velocity/cmd_vel', 10)

        # ── Service clients 
        self._arm_client = self.create_client(
            CommandBool, '/mavros/cmd/arming')
        self._mode_client = self.create_client(
            SetMode, '/mavros/set_mode')

        # ── Heartbeat timer (10 Hz) 
        self.create_timer(0.1, self._heartbeat)

        self.get_logger().info(
            'AerialController ready. '
            'Publish TwistStamped → /drone/cmd_vel | '
            'Bool → /drone/arm')

    # ── Callbacks 

    def _cmd_vel_cb(self, msg: TwistStamped):
        self._latest_cmd = msg
        self._last_cmd_time = self.get_clock().now()
        self._has_cmd = True

    def _state_cb(self, msg: State):
        self._connected = msg.connected
        self._armed = msg.armed
        self._mode = msg.mode

    def _arm_cb(self, msg: Bool):
        self._send_arm(msg.data)

    # ── Heartbeat — streams setpoints at 10 Hz 

    def _heartbeat(self):
        if not self._connected:
            return

        cmd = TwistStamped()
        cmd.header.stamp = self.get_clock().now().to_msg()
        cmd.header.frame_id = 'base_link'

        if self._has_cmd:
            age = (self.get_clock().now() - self._last_cmd_time).nanoseconds / 1e9
            if age < 0.5:
                cmd.twist = self._latest_cmd.twist

        self._vel_pub.publish(cmd)

    # ── Service calls 

    def _send_arm(self, arm: bool):
        if not self._arm_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().warn('Arming service not available')
            return
        req = CommandBool.Request()
        req.value = arm
        future = self._arm_client.call_async(req)
        future.add_done_callback(
            lambda f: self._arm_response(f, arm))

    def _arm_response(self, future, arm: bool):
        try:
            resp = future.result()
            tag = 'ARM' if arm else 'DISARM'
            if resp.success:
                self.get_logger().info(f'{tag} successful')
            else:
                self.get_logger().warn(f'{tag} FAILED (result={resp.result})')
        except Exception as e:
            self.get_logger().error(f'Arm service call failed: {e}')

    def set_mode(self, mode: str):
        """Public helper — used by drone_flight_manager."""
        if not self._mode_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().warn('SetMode service not available')
            return
        req = SetMode.Request()
        req.custom_mode = mode
        future = self._mode_client.call_async(req)
        future.add_done_callback(
            lambda f: self.get_logger().info(
                f'SetMode → {mode}: {"OK" if f.result().mode_sent else "FAIL"}'))


def main(args=None):
    rclpy.init(args=args)
    node = AerialControllerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()
